File: src-web/api/analytics.ts
import { apiClient } from './client';

export type TopQuestionItem = {
  cluster_id: string;
  question: string;
  asks_this_week: number;
};

export type PastQuestionItem = {
  cluster_id: string;
  question: string;
  asks_before_week: number;
  asks_this_week: number;
  asks_total_until_now: number;
};

export type CourseQuestionsAnalyticsResponse = {
  course_id: string;
  timezone: string;
  range_start_utc: string;
  range_end_utc: string;
  top_questions_this_week: TopQuestionItem[];
  past_questions: PastQuestionItem[];
};

export async function fetchCourseQuestionsAnalytics(
  courseId: string,
  rangeStartUtc: string,
  rangeEndUtc: string,
  timezone: string,
) {
  const response = await apiClient.get<CourseQuestionsAnalyticsResponse>('/api/analytics/course-questions', {
    params: {
      course_id: courseId,
      range_start_utc: rangeStartUtc,
      range_end_utc: rangeEndUtc,
      timezone,
    },
  });
  return response.data;
}

File: backend/app/schemas/analytics.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class CourseQuestionsQuery(BaseModel):
    model_config = ConfigDict(extra="forbid")

    course_id: str = Field(min_length=1)
    range_start_utc: str = Field(min_length=1)
    range_end_utc: str = Field(min_length=1)
    timezone: str = Field(default="UTC", min_length=1)
    limit_top: int = Field(default=5, ge=1, le=20)
    limit_past: int = Field(default=5, ge=1, le=30)


class TopQuestionItem(BaseModel):
    cluster_id: str
    question: str
    asks_this_week: int


class PastQuestionItem(BaseModel):
    cluster_id: str
    question: str
    asks_before_week: int
    asks_this_week: int
    asks_total_until_now: int


class ReviewMaterialItem(BaseModel):
    material_id: str
    file_name: str
    page: int
    score: float


class WeeklyImprovementItem(BaseModel):
    cluster_id: str
    question: str
    asks_this_week: int
    asks_before_week: int
    asks_total_until_now: int
    problem: str
    title: str
    solution: str
    review_materials: list[ReviewMaterialItem] = Field(default_factory=list)


class CourseQuestionsAnalyticsResponse(BaseModel):
    course_id: str
    timezone: str
    range_start_utc: str
    range_end_utc: str
    top_questions_this_week: list[TopQuestionItem]
    past_questions: list[PastQuestionItem]
    weekly_improvements: list[WeeklyImprovementItem]
